core/network/src/consensus_gossip.rs

```diff
@@ -216,12 +250,46 @@ impl<B: BlockT> ConsensusGossip<B> {
 	}
 
 	/// Get data of valid, incoming messages for a topic (but might have expired meanwhile)
-	pub fn messages_for(&mut self, topic: B::Hash) -> mpsc::UnboundedReceiver<Vec<u8>> {
+	pub fn messages_for(&mut self, engine_id: ConsensusEngineId, topic: B::Hash)
+		-> mpsc::UnboundedReceiver<Vec<u8>>
+	{
 		let (tx, rx) = mpsc::unbounded();
-		for entry in self.messages.iter().filter(|e| e.topic == topic) {
-			tx.unbounded_send(entry.message.data.clone()).expect("receiver known to be live; qed");
+
+		let validator = match self.validators.get(&engine_id) {
+			None => {
+				self.live_message_sinks.entry((engine_id, topic)).or_default().push(tx);
+				return rx;
+			}
+			Some(v) => v,
+		};
+
+		for entry in self.messages.iter_mut()
+			.filter(|e| e.topic == topic && e.message.engine_id == engine_id)
+		{
+			let live = match entry.status {
+				Status::Live => true,
+				Status::Future => match validator.validate(&entry.message.data) {
```

Comment:
Why call `validate` again here?

Reply:
The goal of `Future` was to allow the initial call to validate to say:
  - we can figure out the topic
  - we don't have enough information to validate yet.

The next assumption is that when you call `messages_for`, you only do so when you do have enough information to validate (because at that point you know the validator keys or other state information). This should lead to another call to `validate` (which in practice will be the last).